File: src/leetcode/fifth/Code_322_CoinChange.java
package leetcode.fifth;

import java.util.Arrays;

/**
 * 零钱兑换 
 * 状态方程定义：dp[i]为兑换金额为i所需的最大硬币数量
 * 状态转移方程：dp[i] = if (min(coin) > target) dp[i] = max_int; else min(dp[i - coin])
 * 初始状态：dp[i] = amount + 1,dp[0]=0
 * 结果：dp[i]
 * 
 * @since 2020-7-17 Friday 22:42 - 23:16
 */
public class Code_322_CoinChange {
    static int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int c : coins) {
                if (i < c) continue;
                dp[i] = Math.min(dp[i], dp[i - c] + 1);
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        System.out.println(coinChange(new int[] { 1, 2, 5 }, 0));
        System.out.println(coinChange(new int[] { 1, 2, 5 }, 11));
        System.out.println(coinChange(new int[] { 2, 5 }, 10));
        System.out.println(coinChange(new int[] { 3 }, 10));
    }
}

File: src/c-leetcode/first/Code_322_CoinChange.c
#include <stdio.h>

int coinChange(int *coins, int coinsLen, int amount)
{
    int dp[amount + 1];
    dp[0] = 0;
    for (int i = 1; i <= amount; i++)
    {
        dp[i] = amount + 1;
    }
    for (int i = 1; i <= amount; i++)
    {
        for (int j = 0; j < coinsLen; j++)
        {
            dp[i] = min(i - coins[j] >= 0 ? dp[i - coins[j]] + 1 : amount + 1, dp[i]);
        }
    }
    return dp[amount] == amount + 1 ? -1 : dp[amount];
}

int min(int i, int j)
{
    return i < j ? i : j;
}

int main(int argc, char **argv)
{
    int coins[] = {1, 2};
    printf("%d", coinChange(&coins, 2, 11));
}